src/OrchardCore/OrchardCore.Infrastructure/Scripting/CommonGeneratorMethods.cs

```diff
@@ -28,26 +36,46 @@ public class CommonGeneratorMethods : IGlobalMethodProvider
     /// Converts a Base64 encoded gzip stream to an uncompressed Base64 string.
     /// See http://www.txtwizard.net/compression.
     /// </summary>
-    private static readonly GlobalMethod _gZip = new()
+    private readonly GlobalMethod _gZip = new()
     {
         Name = "gzip",
         Method = serviceProvider => (Func<string, string>)(encoded =>
         {
-            var bytes = Convert.FromBase64String(encoded);
-            using var gzip = new GZipStream(new MemoryStream(bytes), CompressionMode.Decompress);
+            var neededLength = GetByteArrayLengthFromBase64(encoded);
+
+            Span<byte> bytes = new byte[neededLength];
+
+            if (!Convert.TryFromBase64String(encoded, bytes, out var _))
+            {
+                throw new FormatException("Invalid Base64 string.");
+            }
+
+            using var stream = MemoryStreamFactory.GetStream();
+            stream.Write(bytes);
+            stream.Seek(0, SeekOrigin.Begin);
 
-            var decompressed = new MemoryStream();
+            using var gZip = new GZipStream(stream, CompressionMode.Decompress, leaveOpen: true);
+
+            using var decompressed = MemoryStreamFactory.GetStream();
             var buffer = new byte[1024];
             int nRead;
 
-            while ((nRead = gzip.Read(buffer, 0, buffer.Length)) > 0)
+            while ((nRead = gZip.Read(buffer, 0, buffer.Length)) > 0)
             {
                 decompressed.Write(buffer, 0, nRead);
             }
 
-            return Convert.ToBase64String(decompressed.ToArray());
+            return Convert.ToBase64String(decompressed.GetBuffer(), 0, (int)decompressed.Length);
         }),
     };
 
-    public IEnumerable<GlobalMethod> GetMethods() => new[] { _base64, _html, _gZip };
+    public IEnumerable<GlobalMethod> GetMethods()
+        => new[] { _base64, _html, _gZip };
+
+    // Base64 string length gives us the original byte length. It encodes 3 bytes into 4 characters.
+    // The formula to calculate the number of bytes from the base64 string length is:
+    private static int GetByteArrayLengthFromBase64(string base64String)
+    {
+        return (base64String.Length * 3) / 4;
```

Comment:
NB: it's important to do the multiplication first

![image](https://github.com/user-attachments/assets/b6f88d6e-2952-4d06-b842-39aae8250ac9)